app/plugins/weasyl.ts

```diff
@@ -11,41 +12,128 @@ export class WeasylPlugin extends BaseSitePlugin {
-        pathParts.pop(); //pop off the unneeded hash
-        let id = pathParts.pop();
+        if(canonicalLink.indexOf("/submissions/") > -1) {
+            // regular submission
+            logger.log("weasyl: regular submission");
 
-        // Strip off the username
-        let { filename, ext } = getFilenameParts(serviceFilename);
-        filename = filename.substring(username.length + 1);
+            // Get the username and kill leading '%7E'
+            username = pathParts[1] || "";
+            username = username.slice(3);
+            logger.log("weasyl: username ", username);
 
+            // Record locations of these pieces for later
+            submissionIDIndex = 3;
+            filenameIndex = 4;
+        }
+        else if(canonicalLink.indexOf("/character/") > -1) {
+            // character submission
+            logger.log("weasyl: character submission");
+
+            // Username isn't in canonical link, so get it from the title
+            // above the artwork, which is of the format
+            // <h1 id="detail-title" class="pad-left pad-right">CoolGuy <i>by</i>
+            //   <a class="username" href="/~exampleartist">ExampleArtist</a></h1>
+            let usernameLink = document.querySelector('#detail-title a.username');
+            username = usernameLink.getAttribute("href");
+
+            // Kill leading '/~'
+            username = username.slice(2);
+            logger.log("weasyl: username ", username);
+
+            // Record locations of these pieces for later
+            submissionIDIndex = 2;
+            filenameIndex = 3;
+        }
+        else {
+            // something weird - bail out
+            return Promise.resolve(null);
+        }
+
+        // Get the submission id
+        let id = pathParts[submissionIDIndex] || "";
+        logger.log("weasyl: submission id ", id);
+
+        // Use the title in the canonical link as part of the local filename
+        // Canonical link is silent on the extension (jpg, png, mp3, etc),
+        // so get that later
+        let filename = pathParts[filenameIndex] || "";
+        logger.log("weasyl: filename ", filename);
+        // FIXME  what if filename is blank here?
+
+        // Get extension from the link in the native "download" button.
+        // Don't use the preview URL for this, because it will always be an
+        // image, even for non-image submssions.
+
+        // Get the image URL from the download button, find the last '.' in it,
+        // and take everything after that as the extension
+        let url = button.getAttribute("href");
+        let lastdot = url.lastIndexOf('.');
+        let ext = url.slice(lastdot + 1);
+        logger.log("weasyl: extension ", ext);
+        // FIXME what if ext is blank here?
+
+        // Use the filename from the "canonical" link, plus the extension from
+        // the download button, as serviceFilename
+        // FIXME what does serviceFilename actually do?
```

Comment:
`serviceFilename` is just whatever the website thinks the filename is, as best as raccoony can tell. Usually I grab whatever filename I find in the download button or image or whatever and use that if I can.

In the case of weasyl, given a download button URL of the format `https://cdn.weasyl.com/~[user]/submissions/[id]/[hash]/[serviceFilename.ext]?download`, it'd be the `[serviceFilename.ext]` part.

Reply:
I wrote that comment when I made the very first version of this patch, where I just set `serviceFilename` to a constant string because I hadn't looked closely at what it needed to be.  If you want to kill the FIXME comment, go ahead.  (Maybe also document what each member of I.Media is supposed to be, either where that struct is defined, or on the wiki, or something.)

> it'd be the `[serviceFilename.ext]` part.

I think that's what I end up setting it to.